doc/pact.1
.\"
.\" Source Version: 3.0
.\" Software Release #: LLNL-CODE-422942
.\"
.\" include cpyright.h
.\"

.TH PACT 1 "29 September 2012"
.SH NAME
pact \- a make-type utility featuring persistence and distributed parallel
operation.  Also reports how it was configured.

.SH SYNOPSIS
.B pact [-async <n>] [-B] [-cmd <file> <cmd>] [-d] [-dbg] [-dmp] [-f <file>]
.B      [-flt <file>] [-h] [-i] [-log <file>] [-n] [-na <n>] [+na <n>] 
.B      [-rcr] [-s] [-sys <dir>] [-v]
.B      <make-args> <target>

.B pact [-db] [-incpath] [-info <pttrn> | <var>] [+info <var>] [-link] [+l]

The pact utility fills two roles.  It provides a persistent make facility and
it provides information about the configuration of the PACT distribution
which compiled it.  The latter is very useful for applications that use PACT.
They have an easy way of querying how PACT is configured so that they can
configure themselves appropriately.  Most useful is determining which non-PACT
libraries must also be used to link and the paths to them.

.SH MAKE-MODE OPTIONS

.TP
.B \-async
Issue commands asynchronously to <n> nodes.
.TP
.B \-B
Build Makefile from pre-Make.
.TP
.B \-cmd
Run command <cmd> in directories specified in <file>.
This must come last in the argument list.
.TP
.B \-d
Show all rules and variables.
.TP
.B \-dbg
Attach a debugger when a trapped signal is received.
.TP
.B \-dmp
Dump the internal logs at the end of the session.
.TP
.B \-f
Use <file> for all targets, rules, and variables.
.TP
.B \-flt
Use <file> for specifications to filter out unwanted output.
.TP
.B \-h
A summary of pact options.
.TP
.B \-i
Continue on after errors.
.TP
.B \-log
Write output to <file> in addition to stdout.
.TP
.B \-n
Only print commands, do not execute them.
.TP
.B \-na
Attempt each command at most <n> times before failing.
.TP
.B \+na
Like -na but applies to all child pact sessions too.
.TP
.B \-rcr
Expand simple recursive invocations in place.
.TP
.B \-s
Do not print commands before executing them.
.TP
.B \-sys
Directory in which to put/find Makefile.
.TP
.B \-v
Report the PACT version.
.TP
.B \-vrb
Print all commands before executing them (overrides '@').

.SH INFO-MODE OPTIONS

.TP
.B \-db
Use the database only (default is off).
.TP
.B \-incpath
Report elements needed to compile files.
.TP
.B \-info
Report configuration elements matching the argument.
.TP
.B \+info
Complete report of configuration variable matching the argument.
.TP
.B \-link
Report elements needed to link applications.
.TP
.B \+l
Report only exact matches of the argument.

.SH DESCRIPTION

The PACT project has a system for building code projects in an extremely
high performance fashion.  Extremely high
performance means that from a single set of sources on each
specified network, the code is simultaneously built
for each system type (each build being done in parallel).

This system, called do-net, is a set of C Shell scripts which ultimately
used the version of make provided on each computer system.  It exploits
features of the build system for the code project which is also composed
of shell scripts (or Perl, or Python, ...).  This system works very well
when the networks are relatively unloaded.  As the network/system load
increases we see the effects of lost data packets in the network to the
degree that the failure rate of builds goes to an unacceptable level,
rendering the system less useful than it was designed to be.

The main motivation for the pact utility then is to make the system
highly fault tolerant.  It addresses the following issues:

  1) Lack of persistence in make.
     For one reason or another a command under the control of make
     will fail.  Two important examples are: inability to get a
     compiler license; and temporary inability to access a header
     file.  What a human being does in this case is to retry the
     command.  Very often it works.  We are unaware of any make
     that has this kind of persistence.

     It is often insufficient to simply reissue the make command
     because a partial, corrupt result (e.g., an object file) may
     be left which fools the make dependency analysis into deciding
     that there is no work to be done.  Instead, we need make to
     retry each command until it either succeeds or the number of
     retry attempts exceeds a user specified value.

  2) Buggy make utilities.
     Make utilities have various bugs on different systems.  
     In the case of portable make utilities such as the GNU
     make utility, gmake, they can be corrupted by latent bugs in
     the operating system.  For example, on OSF where child
     processes do not always deliver a SIGCHLD to their parents
     when they exit, gmake sometimes fails and very ungracefully
     at that.

  3) Long process chains of scripts are more failure prone.
     With scripts that invoke other scripts our builds develop
     long chains of processes.  Under high load where dropped
     packets occur, longer chains imply higher probability of
     failure.

  4) Difficulty managing hosts.
     With distributed parallel builds we have to communicate
     with other hosts on the networks we use.  We use SSH for
     this purpose.  The difficulty is that there is a large
     gap between what we need in order to do a distributed
     build and what we can get from the standard system utilities.
     We need to know which hosts are up and which are down.
     We need to know which are heavily loaded and which are only
     lightly loaded.  We need this information quicky and reliably.
     Delays of seconds at a time to get this information and
     use it cannot be tolerated.

     Worse than the latency problem is that a host may not
     respond to a query and also not time out.  With scripts it
     is harder to manage this problem.  With a compiled program,
     the control of timeouts can reside in our own code.

     Technically this is a side issue but it is a significant
     burden which must be overcome, and the solution we have
     developed was designed to address this problem along with
     the rest.

  5) Facility to filter out junk messages.
     Since we are doing many compilations on many hosts it
     is very important to avoid visual clutter that hides
     problems.  Chatty compilers are a nuisance.  The pact
     utility provides means to filter out user specified
     unwanted messages.

The solution to our problem is to jump in and turn our pact utility
into a full blown independent make.  By designing a feature set
to address the problems outlined above we hope to reduce the
set of problems, increase the fault tolerance, and have a utility
that we can modify and debug to handle the most minute issues.

NOTE: because there is a bootstrapping issue inherent in having
a PACT utility that needs PACT to be built in order to build itself,
we have two utilies: smake and dmake.  Smake is a simple wrapper
around the system make (hence the name smake).  We use smake just
long enough to build our full blown utility called dmake and let
it do the rest. Then pact is a link to dmake.  This scheme provides
maximum flexibility in the specification of our pre-Make files and
gives us options when porting to new platforms (i.e. use smake until
dmake is operational).  In the remainder of this document pact and
dmake may be used interchangeably.


.SH FEATURES

The features built into dmake include:

  1) Persistence
     Attempts commands a specified number of times (3 by default).

  2) Parallel
     Using a facility (documented elsewhere) referred to as
     the host-server, the set of commands obtained from the
     dependency analysis are executed asynchronously on a
     set of nodes matching a specific system type.  This can be
     done efficiently at either fine or coarse granularity.  That
     is to say, you can dispatch individual compilations or whole
     subsystems with very low latency.

     A BARRIER specification is added so that groups of commands
     can be forced to finish before others are dispatched.

  3) Client/Server Model

     When dmake is run with the -async option it computes the number
     of servers it will need to execute the given commands.  It will
     then launch each server on a different host (supplied by the
     host-server).  Each server will be sent requests to run commands,
     it will run them, finally it will return the output and the exit
     status of each job to the dmake client session.  The servers
     persist until the client terminates them at the end of the
     session.

     The client and servers monitor commands that are run.  In an
     effort to avoid commands that hang or are lost by the system
     causing an entire session to fail, time limits are placed on
     them.  If a command does not complete within the alloted time
     it is killed and retried.  There are several controls that
     determine the time limits for a command, but an important
     factor in determining how long to wait is how long other
     commands executed by the same server take.  So the client
     records how long each command takes to run and calculates
     an average for each server.

     For diagnostic, debugging, and system monitoring reasons the
     client keeps a log of all traffic with each server and each
     server keeps a log of all traffic with each job as well as
     the client.  The client also keeps a record (dubbed stats) of
     the load average, free memory, average job time to completion,
     NFS traffic, and job assignments for each server.

     A server process is invoked as 'dmake -srv'.  Note the -srv
     option is otherwise undocumented and should never be used.

  4) Host Management
     We have made the attempt to abstract out the use of hosts.
     A combination of hardware and OS is referred to as a system
     type. A configuration includes the specification of the system
     type and the compilers.  When building a configuration any
     host of the correct system type that mounts the file system
     where the sources are is as good as any other for doing the
     work.  Using a sub-system called the host server which
     maintains a list of hosts associated with a system type, we
     can build for a system type rather than building on a
     specific host.

     With this abstraction we can use the host server to supply
     hosts that are up and running to execute commands rather
     than making requests to specific hosts which might be down.
     This model gives dmake a great deal of robustness while
     executing commands asynchronously.

  5) Library
     The functionality of dmake is bound in a set of SCORE 
     library calls. Figuratively they are:

         state = SC_make_state()
         SC_make_def_var(s, state)
         SC_make_def_rule(s, state)
         SC_parse_makefile(fname, state)
         SC_parse_premake(fname, state)
         SC_analyze_dependencies(target, state)
         cmnds = SC_action_commands(state)
         SC_free_state(state)
         
         SC_exec_commands(..., cmnds, ...)
         SC_exec_async(..., cmnds, ...)

     This approach allows you to build make-like capabilities into
     your own application or to build a tailored make type utility.

  6) Recursion
     Simple recursions may be expanded in situ in order to reduce the
     number of processes used.  This is controlled by the -rcr option.

     For example, the following rule:

         install:
              pact link
              pact incinstall

     would normally cause two commands to be issued: 'pact link' and
     'pact incinstall'.  With recursion the original session simply
     does the dependency analysis for each of the targets 'link' and
     'incinstall' and inserts any resulting commands into the command
     set in place of the pact invocation commands.


manager/news
6/5/2012

For a long time the PACT Python bindings
were made using the Python Numeric package.  This has been
superceeded by the NumPy package.  When we build PACT
we look for both Numeric and NumPy now.  Because the two
packages have some differences we build a set of bindings for
each one that is available.  Since Numeric is the original package,
the bindings using it are in:

     /usr/local/apps/pactnew/gnu/lib/pythond.d/pact

and the new Numeric bindings are in:

     /usr/local/apps/pactnew/gnu/lib/pythond.d/pact_2

Each of those directories will be there if and only if the
Python package that it corresponds to is there.

The Chaos 5 machines appear to have only the NumPy package
and consequently only the pact_2 bindings.

To use the Numeric bindings you do:

    setenv PYTHONPATH  /usr/local/apps/pactnew/gnu/lib/pythond.d/pact

    and

    import pact.pdb as pdb

To use the NumPy bindings you do:

    setenv PYTHONPATH  /usr/local/apps/pactnew/gnu/lib/pythond.d/pact_2

    and

    import pact_2.pdb as pdb


1/26/2011

PDBLib supports C99 data types (version 26).

SCORE type manager expanded and made integral to
PACT in general.  The template utility was added to
generate functions for specified data types.  The
advantage over macros is debuggability and more
nearly first class behavior with respect to types.

More general polygon intersection handles non-simple,
non-convex polygons, multiple discreet polygon
intersections, and polygon unions as well.


11/10/2009

For developers, moved and consolidated "most"
#define'd constants into named enum's.  This allows the
compiler to do better error checking and helps in debugging
where you see symbolic names rather than numerical values.


For end users, there are many, many new features:

1) ULTRA now has a fill command to allow you to color
   the area under curves or fill closed curves.

2) Colors can be expressed as RGB values in many
   contexts.  I intend to get them all in time.
   An RGB color is specified as:

      #x7frrggbb

   where r, g, and b above are hex digits.

3) If you do the following in your .ultrarc

   ld mouse.scm

   you can can control the viewport with the
   mouse.  The functionality is more or less modelled
   after the Google map and other similar usages.

4) PDBView can now look at PDB files inside TAR
   files or AR archives.  This currently works only
   with a serial version of PDBView

5) PDBView can read mySQL and SQLite databases.
   Writing capabilities will come next.  This capability
   depends on the appropriate libraries and headers
   being installed on the platform where PACT is built.

6) A rudimentary ability to write PDB files which are
   plain text.  This will be fully developed in future
   releases.

01/10/2008

Collect PGS global attribute variables into struct and add
PG_get_attrs_glb, PG_set_attrs_glb, and PG_ptr_attrs_glb as the
sole accessors.  This keeps PGS state from conflicting with
applications or forcing them to change unnecessarily.

11/18/2007

Divide headers into public and private parts.  The publics parts
have the old names (e.g. score.h).  The private parts have the
old name with _int added (e.g. score_int.h).  Move most of the
_XX variables and functions into the private headers to hide them
from applications which should use standard accessors in the
public parts.


04/19/2006

Introduce PDB format version III.


08/13/2004

Introduce the hserve utility.


trunk/source/manager/pact.c
/*
 * PACT.C - wrapper to control make to PACT's needs
 *
 * Source Version: 3.0
 * Software Release #: LLNL-CODE-422942
 *
 * include "cpyright.h"
 *
 */

#include <scconfig.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#define MAXLINE 4096
#define TRUE    1
#define FALSE   0

#undef max
#define max(a, b) (a > b) ? a : b
#undef min
#define min(a, b) (a < b) ? a : b

static int
 show = 0;

static char
 sys[MAXLINE],
 arch[MAXLINE],
 root[MAXLINE],
 cwd[MAXLINE],
 srcdir[MAXLINE],
 tmpdir[MAXLINE];

int system();
int atoi();
char *getcwd();
char *getenv();
void (*signal())();
void exit(int status);

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* NSTRCAT - safe strcat function */

static char *nstrcat(char *d, int nc, char *s)
   {int n;

    n = nc - 1 - strlen(d);
    n = min(n, strlen(s));

    strncat(d, s, n);

    return(d);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* REPORT_VAR - report on the variable Q in FILE */

static int report_var(char *file, char *q, char *key,
		      int compl, int litrl, int newl)
   {int i, ok, doit, tst;
    char s[MAXLINE];
    char *tok, *txt, *ps;
    FILE *fp;

    ok = FALSE;

    fp = fopen(file, "r");
    if (fp != NULL)
       {for (i = 0; fgets(s, MAXLINE, fp) != NULL; i++)
	    {if (key != NULL)
	        {tok  = strtok(s, " \t\r");
		 doit = ((tok != NULL) && (strcmp(tok, key) == 0));
		 ps   = NULL;}
	     else
	        {doit = TRUE;
		 ps   = s;};

	     if (doit)
	        {tok = strtok(ps, " \t\n");
		 if (tok == NULL)
		    continue;

		 tst = (litrl) ? (strcmp(tok, q) == 0) :
		                 (strncmp(tok, q, strlen(q)) == 0);

		 if (tst)
		    {txt = strtok(NULL, "\n");
		     if (txt == NULL)
		        printf("%s", tok);

		     else
		        {while (*txt != '\0')
			    {if (strchr("= \t", *txt) == NULL)
			        break;
			     else
			        txt++;};

			 if (compl)
			    printf("%s = %s", tok, txt);
			 else
			    printf("%s", txt);

			 memset(s, 0, MAXLINE);
			 ok = TRUE;};

		     if (newl == TRUE)
		        printf("\n");
                     else
		        printf(" ");};};};

	fclose(fp);};

    return(ok);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* REPORT - report the value of one of the configuration quantities */

static void report(char *q, int compl, int litrl, int newl)
   {int ok;
    char devdir[MAXLINE], file[MAXLINE];
    char *s;

    strcpy(devdir, root);

    ok = FALSE;

    if (strcmp(q, "make") == 0)
       {snprintf(file, MAXLINE, "%s/include/make-def", devdir);
        ok = report_var(file, "UMake", NULL, compl, litrl, newl);};

    if (strcmp(q, "config") == 0)
       {snprintf(file, MAXLINE, "%s/include/make-def", devdir);
        ok = report_var(file, "System", NULL, compl, litrl, newl);};

    if (!ok)
       {snprintf(file, MAXLINE, "%s/include/scconfig.h", devdir);
	ok = report_var(file, q, "#define", compl, litrl, newl);};

    if (!ok)
       {snprintf(file, MAXLINE, "%s/include/make-def", devdir);
	ok = report_var(file, q, NULL, compl, litrl, newl);};

    if (!ok)
       {snprintf(file, MAXLINE, "%s/include/configured", devdir);
	ok = report_var(file, q, NULL, compl, litrl, newl);};

    if (!ok)
       {s = getenv("SHELL");
	if (s != NULL)
	   {if (strstr(s, "csh") != NULL)
	       {snprintf(file, MAXLINE, "%s/include/env-pact.csh", devdir);
		ok = report_var(file, q, "setenv", compl, litrl, newl);}
	    else
	       {snprintf(file, MAXLINE, "%s/include/env-pact.sh", devdir);
		ok = report_var(file, q, "export", compl, litrl, newl);};};};

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* REPORT_CL - report the information relevant to compiling and linking */

static void report_cl(char *q)
   {

    if (strcmp(q, "-incpath") == 0)
       {report("MDInc", FALSE, TRUE, FALSE);
	report("MDGInc", FALSE, TRUE, TRUE);}

    else if (strcmp(q, "-link") == 0)
       {report("MDGLib", FALSE, TRUE, FALSE);
	report("DPLib", FALSE, TRUE, FALSE);
	report("MDLib", FALSE, TRUE, TRUE);};

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* MANAGE_TMP_DIR - manage a temporary directory to guarantee multi-platform
 *                - builds do not interfere with each other
 *                - the DIR_WAS_THERE variable is to manage the situation
 *                - of recursive pact invocations
 */

void manage_tmp_dir(char *cwd, char *sname, int start)
   {int ss;
    char cmd[MAXLINE];
    static int dir_was_there = FALSE;

/* do this at the beginning */
    if (start == TRUE)
       {strcpy(srcdir, cwd);

#ifdef USE_TMP_DIR
        if (strncmp(cwd, "/tmp/pact_make_", 15) == 0)
	   strcpy(tmpdir, cwd);
        else
	    sprintf(tmpdir, "/tmp/pact_make_%d", (int) getpid());
#else
	sprintf(tmpdir, "%s/z-%s/obj", cwd, sname);
#endif

	snprintf(cmd, MAXLINE, "test -d %s/", tmpdir);
	ss = system(cmd);
	if (ss != 0)
	   dir_was_there = FALSE;
	else
	   dir_was_there = TRUE;

        if (!dir_was_there)
	   {snprintf(cmd, MAXLINE, "mkdir -p %s/", tmpdir);
	    ss = system(cmd);};}

/* do this at the end */
    else
       {if (!dir_was_there)
	   {snprintf(cmd, MAXLINE, "rm -rf %s/", tmpdir);
	    ss = system(cmd);};};


    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* BUILD_MAKEFILE - make the specified Makefile from the pre-Make
 *                - in the current directory
 *                - return 0 iff successful
 */

int build_makefile(char *mkfile)
   {int err;
    char cmd[MAXLINE], makef[MAXLINE], inc[MAXLINE];

    err = 0;

    snprintf(inc, MAXLINE, "%s/include", root);
    snprintf(makef, MAXLINE, "%s", mkfile);

    snprintf(cmd, MAXLINE, "csh -c \"mkdir -p %s/obj >& /dev/null\"", arch);
    system(cmd);

/*    printf("\nMaking %s from pre-Make\n\n", makef); */

    snprintf(cmd, MAXLINE, "cp %s/make-def %s", inc, makef);
    err |= system(cmd);

    snprintf(cmd, MAXLINE, "echo PACTTmpDir = %s/obj >> %s", arch, makef);
    err |= system(cmd);

    snprintf(cmd, MAXLINE, "echo PACTSrcDir = ../.. >> %s", makef);
    err |= system(cmd);

    snprintf(cmd, MAXLINE, "cat pre-Make >> %s", makef);
    err |= system(cmd);

    snprintf(cmd, MAXLINE, "cat %s/make-macros >> %s", inc, makef);
    err |= system(cmd);

    return(err);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* HANDLER - handle signals gracefully */

static void handler(int sig)
   {

    fprintf(stdout, "PACT exiting with signal %d\n", sig);

    exit(sig);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* INVOKE_MAKE - complete the command line CMD and exec it */

static int invoke_make(char *cmd, int nc, int c, char **v)
   {int i, st, na, ns;
    char s[MAXLINE];
    char *log;
    
#ifdef USE_GNU_MAKE
    snprintf(s, MAXLINE, "gmake --no-print-directory -f -");
#else
    snprintf(s, MAXLINE, "make -f -");
#endif

    nstrcat(cmd, nc, s);

    na  = 1;
    log = NULL;
    for (i = 1; i < c; i++)
        {if (strcmp(v[i], "-f") == 0)
            i++;

         else if (strcmp(v[i], "-log") == 0)
            log = v[++i];

	 else if ((strcmp(v[i], "-na") == 0)   ||
		  (strcmp(v[i], "+na") == 0))
            na = atoi(v[++i]);

	 else if (strcmp(v[i], "-async") == 0)
	    {

#ifdef USE_GNU_MAKE
	     snprintf(s, MAXLINE, " -j %d", atoi(v[++i]));
	     nstrcat(cmd, nc, s);
#endif
	    }

	 else
	    {nstrcat(cmd, nc, " ");
	     nstrcat(cmd, nc, v[i]);};};

    if (log != NULL)
       {snprintf(s, MAXLINE, " | tee -a %s", log);
	nstrcat(cmd, nc, s);};

/* do the make */
    if (show)
       printf("Command: %s\n", cmd);

/* retry until successful */
    for (i = 0; i < na; i++)
        {st = system(cmd);   
	 if (st == 0)
	    break;
	 else
            {ns = (i < 1) ? 1 : 30;
	     printf("***> failed (%d) - attempt %d in %d seconds\n",
		    st, i+2, ns);
	     printf("***>      %s\n", cmd);
	     printf("***> retry '%s' - task 1\n", cmd);
	     sleep(ns);};};

    return(st);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* METHOD_1 - try to set make-def, pre-Make, and make-macros to make
 *          - via stdin
 *          - PROBLEM: on some systems the input to stdin is routed
 *          - to a temporary file and file system errors under load
 *          - can kill you
 */

static int method_1(int c, char **v, char *pmname)
   {int i, status;
    char cmd[MAXLINE], s[MAXLINE];

/* find the current directory */
    if (getcwd(cwd, MAXLINE) == NULL)
       {fprintf(stderr, "ERROR: CAN'T GET CURRENT DIRECTORY\n");
        return(1);};

/* manage the hidden directory for the temporary files */
    manage_tmp_dir(cwd, sys, TRUE);

/* write the command line */
    strcpy(cmd, "(");
	
    i = strlen(cwd) - 7;
    i = max(i, 0);
    if (strcmp(cwd+i, "manager") == 0)
       {snprintf(s, MAXLINE, "echo \"System = %s\" ; ", sys);

        snprintf(s, MAXLINE, "cat %s", pmname);
	nstrcat(cmd, MAXLINE, s);

	nstrcat(cmd, MAXLINE, ") | ");}

    else
       {snprintf(s, MAXLINE, "echo \"include %s/include/make-def\" ; ", root);
	nstrcat(cmd, MAXLINE, s);

        snprintf(s, MAXLINE, "echo \"PACTTmpDir = %s\" ; ", tmpdir);
	nstrcat(cmd, MAXLINE, s);

        snprintf(s, MAXLINE, "echo \"PACTSrcDir = %s\" ; ", srcdir);
	nstrcat(cmd, MAXLINE, s);

        snprintf(s, MAXLINE, "cat %s ; ", pmname);
	nstrcat(cmd, MAXLINE, s);

	snprintf(s, MAXLINE, "echo \"include %s/include/make-macros\"", root);
	nstrcat(cmd, MAXLINE, s);

	nstrcat(cmd, MAXLINE, ") | ");};

/* make the command to invoke make */
    status = invoke_make(cmd, MAXLINE, c, v);

/* remove the hidden directory for the temporary files */
    manage_tmp_dir(cwd, sys, FALSE);

    return(status);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* METHOD_2 - try to use a pre-built Makefile in MKFILE */

static int method_2(int c, char **v, char *mkfile, long tmm)
   {int status;
    long tmp;
    char cmd[MAXLINE];
    struct stat st;

/* check to see if the pre-Make is newer than the Makefile */
    if (stat("pre-Make", &st) == 0)
       {if (S_ISREG(st.st_mode))
	   {tmp = st.st_mtime;
	    if (tmp > tmm)
               build_makefile(mkfile);};};

/* make the command to invoke make */
    cmd[0] = '\0';
    status = invoke_make(cmd, MAXLINE, c, v);

    return(status);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* POP_PATH - pop the tail of the path off
 *          - return a pointer to the head component
 */

char *pop_path(char *path)
   {char *p;

    p = strrchr(path, '/');
    if (p != NULL)
       *p++ = '\0';
    else
       p = path;

    return(p);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* IS_EXECUTABLE_FILE - if PATH is the name of an executable file
 *                    - and if the length of path is <= NCX
 *                    - copy it into FP and return the length
 */

static int is_executable_file(char *fp, char *path, int ncx)
   {int n, muid, mgid, fuid, fgid;
    int usrx, grpx, othx, file;
    struct stat bf;

    n = -1;

    muid = getuid();
    mgid = getgid();

    n = -1;
    if (stat(path, &bf) == 0)
       {fuid = bf.st_uid;
	fgid = bf.st_gid;
	file = bf.st_mode & S_IFREG;
	usrx = ((muid == fuid) && (bf.st_mode & S_IXUSR));
	grpx = ((mgid == fgid) && (bf.st_mode & S_IXGRP));
	othx = (bf.st_mode & S_IXOTH);
	if (file && (usrx || grpx || othx))
	   {n = strlen(path);
	    if (n <= ncx)
	       {strcpy(fp, path);
		n = 0;};};}

    return(n);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* HANDLE_PATH_DOT - given a directory DIR and path NAME
 *                 - resolve out ./ and ../ elements of NAME
 *                 - return result in PATH
 *                 - DIR may be destroyed in this process
 */

static void handle_path_dot(char *path, char *dir, char *name)
   {char *s;

    if (dir != NULL)
       {s = name;

/* loop over any number of ./ or ../ elements of the path */
	while (s[0] == '.')
	   {if (s[1] == '.')
	       {s += 3;
		pop_path(dir);}
	    else
	       s += 2;};

/* construct the candidate path */
	sprintf(path, "%s/%s", dir, s);};

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* FILE_PATH - given the name of a file return the full path
 *           - if FULL is TRUE
 *           - otherwise expand out '.' and '..' constructions
 *           - input is the NAME, output is the PATH which is
 *           - at most NC chars long
 *           - return:
 *           -    0 on success
 *           -    n NAME is there but NC is too few characters
 *           -      n is the number of characters needed to contain
 *           -      the full path
 */

int file_path(char *name, char *path, int nc)
   {int n;
    size_t nb;
    char pathvar[MAXLINE], fp[MAXLINE];
    char *t, *p;
    extern char *getcwd(char *buf, size_t size);

    n  = -1;
    nb = MAXLINE - 1;
    switch (name[0])
       {case '/' :
	     n = is_executable_file(path, name, nc);
             break;

        case '.' :
             t = getcwd(pathvar, nb);
             handle_path_dot(fp, t, name);
	     n = is_executable_file(path, fp, nc);
             break;

        default:
	     p = getenv("PATH");
	     if (p != NULL)
	        {strncpy(pathvar, p, MAXLINE);
		 pathvar[MAXLINE-1] = '\0';

		 for (t = strtok(pathvar, ":");
		      t != NULL;
		      t = strtok(NULL, ":"))
		     {handle_path_dot(fp, t, name);
		      n = is_executable_file(path, fp, nc);
		      if (n == 0)
			 break;};};
             break;};

    return(n);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SETUP_ENV - initialize the ARCH and ROOT strings
 *           - return TRUE iff successful with both
 */

static int setup_env(char *src, char *arch, char *root)
   {char s[MAXLINE];
    char *p, *ps;

    file_path(src, root, MAXLINE);

    sys[0] = '\0';

/* remove exe name */
    p = pop_path(root);

/* remove bin dir */
    p = pop_path(root);

/* setup the architecture string */
    p = getenv("SESSION_CONFIG");
    if (p != NULL)
       {strncpy(sys, p, MAXLINE);
	sys[MAXLINE-1] = '\0';}

    else
       {strcpy(s, root);

	ps = pop_path(s);
	p  = pop_path(s);

/* if root is of the form /.../dev/<arch> use <arch> */
	if (strcmp(p, "dev") == 0)
	   strcpy(sys, ps);

/* otherwise use the configured system_id */
	else
	   strcpy(sys, SYSTEM_ID);};

    sprintf(arch, "z-%s", sys);

    return(TRUE);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* USAGE - print help */

void usage(void)
   {

    printf("\n");
    printf("Usage: pact [-async <n>] [-B] [-h] [-incpath] [-info <pttrn> | <var>] [+info <var>]\n");
    printf("            [-log <file>] [-link] [+l] [-na <n>] [-sys <dir>] [-v] <make-args> <target>\n");
    printf("\n");

    printf("Info Options:\n");
    printf("    -async   number of processes to use\n");
    printf("    -na      number of attempts to make\n");
    printf("    -log     log output to designated file\n");
    printf("\n");

    printf("Info Options:\n");
    printf("    -B       build Makefile from pre-Make\n");
    printf("    -h       this help message\n");
    printf("    -incpath report elements needed to compile files\n");
    printf("    -info    report configuration elements matching the argument\n");
    printf("    +info    complete report of configuration variable matching the argument\n");
    printf("    -link    report elements needed to link applications\n");
    printf("    +l       report only exact matches of the argument\n");
    printf("    -sys     directory in which to put/find Makefile\n");
    printf("    -v       report the PACT version\n");
    printf("\n");
    printf("    All other arguments are passed into make\n");
    printf("\n");

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* MAIN - start here */

int main(int c, char **v)
   {int i, status, keep, ok, nproc, na, litrl, mc;
    long tmm;
    char mkfile[MAXLINE];
    char *pmname, *mv[100];
    static char *default_pmname = "pre-Make";
    struct stat st;

    signal(2, handler);

    pmname = default_pmname;
    keep   = 0;
    na     = 1;
    nproc  = 1;
    show   = 0;

    setup_env(v[0], arch, root);

    ok    = TRUE;
    litrl = FALSE;
    mc    = 1;
    for (i = 1; (i < c) && ok; i++)
        {if (strcmp(v[i], "-sys") == 0)
	    sprintf(arch, "%s", v[++i]);

         else if (strcmp(v[i], "-show") == 0)
	    show = 1;

/* get dmake options taking the next argument later */
	 else if ((strcmp(v[i], "-async") == 0) ||
		  (strcmp(v[i], "-na") == 0)   ||
		  (strcmp(v[i], "+na") == 0)   ||
		  (strcmp(v[i], "-log") == 0))
	    {mv[mc++] = v[i++];
	     mv[mc++] = v[i];}

/* ignore dmake options taking the next argument */
	 else if ((strcmp(v[i], "-dst") == 0)   ||
		  (strcmp(v[i], "-flt") == 0)   ||
		  (strcmp(v[i], "-cmd") == 0))
            {printf("smake: option %s ignored\n", v[i]);
	     i++;}

	 else if (strcmp(v[i], "-incpath") == 0)
	    {report_cl(v[i]);
	     return(0);}

	 else if (strcmp(v[i], "-link") == 0)
	    {report_cl(v[i]);
	     return(0);}

/* ignore dmake options taking no argument */
	 else if ((strcmp(v[i], "-dbg") == 0) ||
		  (strcmp(v[i], "-dmp") == 0) ||
		  (strcmp(v[i], "-srv") == 0) ||
		  (strcmp(v[i], "-vrb") == 0))
            {printf("smake: option %s ignored\n", v[i]);
	     continue;}

         else if (v[i][0] == '-')
            {switch (v[i][1])
                {case 'B' :
                      snprintf(mkfile, MAXLINE, "%s/Makefile", arch);
                      status = build_makefile(mkfile);
		      return(status);
		      break;
		 case 'h' :
		      usage();
		      return(1);
		      break;
		 case 'i' :
		      if (strcmp(v[i], "-info") == 0)
			 {if (++i < c)
                             report(v[i], FALSE, litrl, TRUE);
			  return(0);};
		      i--;
		      ok = FALSE;
		      break;
		 case 'f' :
		      pmname = v[++i];
		      break;
	         case 'n' :
		      keep = 1;
                      mv[mc++] = v[i];
		      break;
	         case 'v' :
		      printf("%s\n", PACT_VERSION);
		      exit(0);
		      break;};}
	 else if (v[i][0] == '+')
            {switch (v[i][1])
                {case 'i' :
		      if (strcmp(v[i], "+info") == 0)
			 {if (++i < c)
                             report(v[i], TRUE, litrl, TRUE);
			  return(0);};
		      break;
		 case 'l' :
		      litrl = TRUE;
		      break;};}
	 else
	    {mv[mc++] = v[i];
	     ok = FALSE;};};

    for (; i < c; i++)
        mv[mc++] = v[i];

    mv[mc] = NULL;
    mv[0]  = v[0];

    snprintf(mkfile, MAXLINE, "%s/Makefile", arch);
    if (stat(mkfile, &st) == 0)
       {if (S_ISREG(st.st_mode))
	   {tmm    = (long) st.st_mtime;
	    status = method_2(mc, mv, mkfile, tmm);
	    return((status == 0) ? 0 : 1);};};

    status = method_1(mc, mv, pmname);

    return((status == 0) ? 0 : 1);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/
